plugins/kotlin-as-java/src/main/kotlin/converters/KotlinToJavaConverter.kt

```diff
@@ -259,24 +262,51 @@ internal fun DClass.asJava(): DClass = copy(
             )
         }, // name may not always be valid here, however classNames should always be not null
     functions = functionsInJava(),
-    properties = properties
-        .filterNot { it.hasJvmSynthetic() }
-        .map { it.asJava() },
-    classlikes = classlikes.map { it.asJava() },
+    properties = propertiesInJava(),
+    classlikes = classlikesInJava(),
     generics = generics.map { it.asJava() },
+    companion = companion.companionAsJava(),
     supertypes = supertypes.mapValues { it.value.map { it.asJava() } },
     modifier = if (modifier.all { (_, v) -> v is KotlinModifier.Empty }) sourceSets.associateWith { JavaModifier.Final }
     else sourceSets.associateWith { modifier.values.first() }
 )
 
+/**
+ * Companion objects requires some custom logic for rendering as Java.
+ * They are excluded from usual classlikes rendering and added after.
+ */
+internal fun DClass.classlikesInJava(): List<DClasslike> {
+    val companionAsJava = companion.companionAsJava()
+
+    val classlikes = classlikes
+        .filter { it.name != companion?.name }
+        .map { it.asJava() }
+    return if (companionAsJava != null) classlikes.plus(companionAsJava) else classlikes
```

Comment:
also nitpick: I'd move `companionAsJava` closer to where it's used, otherwise it makes you think that it's used for something else (i.e for computing `classlikes` val)
```suggestion
    val classlikes = classlikes
        .filter { it.name != companion?.name }
        .map { it.asJava() }
        
    val companionAsJava = companion.companionAsJava()
    return if (companionAsJava != null) classlikes.plus(companionAsJava) else classlikes
```